### DISCUSSION TITLE
Structured Logging

### DISCUSSION THREAD

**Discussion by shaun-nx** [nginx/kubernetes-ingress#6436] [Ideas]
Hi folks 👋

We're using this forum to gauge interest around updating the logging library that NGINX Ingress Controller uses.

Currently, we use [golang/glog](https://pkg.go.dev/github.com/golang/glog).

If we were to move to a logging library that supports structured logging, we would likely choose [log/slog](https://pkg.go.dev/log/slog) mainly to reduce external package dependency.

### Why make this change?

There are a few reason we've been considering this change for the project.

1. Greater performance at massive scale
2. Standardised log format
3. Improved security
4. Remove issues related to leaking goroutines

That being said, we would love to know what you think!
Have you worked on a project that implements a structured logging library?
Did you notice the difference? 
Were there complications that you hit?
We'd love to hear any and all feedback! 😃

**Comment by jjngx** [comment]
Structured logging approach, in contrary to the free-text format currently used, introduces a contract between NIC and log messages consumers. It would help users to parse log messages, aggregate and process them further (eg. Splunk).

**Comment by jjngx** [comment]
error msg:

```shell
goleak: Errors on successful test run: found unexpected goroutines:
[Goroutine 18 in state chan receive, with github.com/golang/glog.(*loggingT).flushDaemon on top of the stack:
github.com/golang/glog.(*loggingT).flushDaemon(0x103d29fc0)
	/Users/j.jarosz/go/pkg/mod/github.com/nginxinc/glog@v1.1.2/glog.go:892 +0x74
created by github.com/golang/glog.init.0 in goroutine 1
	/Users/j.jarosz/go/pkg/mod/github.com/nginxinc/glog@v1.1.2/glog.go:413 +0x2d0
]
FAIL	github.com/nginxinc/kubernetes-ingress/internal/configs	1.093s
```

**Comment by shaun-nx** [comment]
FYI, this is now available in [v4.0.0!](https://github.com/nginx/kubernetes-ingress/releases/tag/v4.0.0)
